In [90]:
import numpy as np 
import networkx as nx

from link_prediction import generate_non_edges, JaccardCoefficient, CommonNeighbors, AdamicAdar

def remove_random_edges(g, fraction):
    mask = np.random.randint(0, 100, len(g.edges())) < fraction * 100
    e_removed = np.array(g.edges())[mask]
    g_train = g.copy()
    g_train.remove_edges_from(e_removed)
    return g_train, e_removed

gfb = nx.read_gml('data/Johns Hopkins55.gml')

fractions = [0.05, 0.1, 0.15, 0.2]
k_values = [50, 100, 200, 400]

f, k = 0.1, 50
def generate_precision(f, k):
    g_train, e_removed = remove_random_edges(gfb, f)
    non_edges = generate_non_edges(g_train)

    jp = JaccardCoefficient(g_train)
    predictions_jp = jp.predict(non_edges)
    cn = CommonNeighbors(g_train)
    predictions_cn = cn.predict(non_edges)
    aa = AdamicAdar(g_train)
    predictions_aa = aa.predict(non_edges)
    precisions = []
    for p in [predictions_jp, predictions_cn, predictions_aa]: 
        predictions_sorted = np.argsort(p)[::-1][:k]
        e_removed_tuples = [(u, v) for u, v in zip(e_removed[:,0], e_removed[:,1])]
        
        tp = 0
        fp = 0
        for x in predictions_sorted:
            if non_edges[x] in e_removed_tuples:
                tp += 1
            else:
                fp += 1
        
        precisions.append(tp / (tp + fp))
    return precisions


In [91]:
import itertools
for x, y in itertools.product(fractions, k_values):
    print(f"Précision pour f = {x}, k = {y} : {generate_precision(x, y)}")

Précision pour f = 0.05, k = 50 : [0.12, 0.5, 0.48]
Précision pour f = 0.05, k = 100 : [0.33, 0.56, 0.52]
Précision pour f = 0.05, k = 200 : [0.34, 0.55, 0.53]
Précision pour f = 0.05, k = 400 : [0.335, 0.4975, 0.5225]
Précision pour f = 0.1, k = 50 : [0.14, 0.76, 0.74]
Précision pour f = 0.1, k = 100 : [0.42, 0.72, 0.73]
Précision pour f = 0.1, k = 200 : [0.51, 0.73, 0.72]
Précision pour f = 0.1, k = 400 : [0.4375, 0.685, 0.6775]
Précision pour f = 0.15, k = 50 : [0.16, 0.86, 0.8]
Précision pour f = 0.15, k = 100 : [0.47, 0.8, 0.76]
Précision pour f = 0.15, k = 200 : [0.54, 0.835, 0.81]
Précision pour f = 0.15, k = 400 : [0.5225, 0.7675, 0.7475]
Précision pour f = 0.2, k = 50 : [0.24, 0.86, 0.84]
Précision pour f = 0.2, k = 100 : [0.48, 0.89, 0.83]
Précision pour f = 0.2, k = 200 : [0.57, 0.84, 0.84]
Précision pour f = 0.2, k = 400 : [0.535, 0.8475, 0.8375]


## Résultats pour Caltech
Précision pour f = 0.05, k = 50 : [0.3, 0.46, 0.46]
Précision pour f = 0.05, k = 100 : [0.32, 0.45, 0.47]
Précision pour f = 0.05, k = 200 : [0.25, 0.35, 0.365]
Précision pour f = 0.05, k = 400 : [0.2025, 0.23, 0.245]
Précision pour f = 0.1, k = 50 : [0.38, 0.7, 0.72]
Précision pour f = 0.1, k = 100 : [0.36, 0.6, 0.59]
Précision pour f = 0.1, k = 200 : [0.345, 0.525, 0.535]
Précision pour f = 0.1, k = 400 : [0.37, 0.45, 0.4625]
Précision pour f = 0.15, k = 50 : [0.42, 0.78, 0.82]
Précision pour f = 0.15, k = 100 : [0.52, 0.79, 0.76]
Précision pour f = 0.15, k = 200 : [0.46, 0.665, 0.665]
Précision pour f = 0.15, k = 400 : [0.4875, 0.5475, 0.555]
Précision pour f = 0.2, k = 50 : [0.42, 0.88, 0.88]
Précision pour f = 0.2, k = 100 : [0.55, 0.84, 0.84]
Précision pour f = 0.2, k = 200 : [0.505, 0.715, 0.72]
Précision pour f = 0.2, k = 400 : [0.5175, 0.68, 0.685]

## Résultats pour Johns Hopkins
Précision pour f = 0.05, k = 50 : [0.12, 0.5, 0.48]
Précision pour f = 0.05, k = 100 : [0.33, 0.56, 0.52]
Précision pour f = 0.05, k = 200 : [0.34, 0.55, 0.53]
Précision pour f = 0.05, k = 400 : [0.335, 0.4975, 0.5225]
Précision pour f = 0.1, k = 50 : [0.14, 0.76, 0.74]
Précision pour f = 0.1, k = 100 : [0.42, 0.72, 0.73]
Précision pour f = 0.1, k = 200 : [0.51, 0.73, 0.72]
Précision pour f = 0.1, k = 400 : [0.4375, 0.685, 0.6775]
Précision pour f = 0.15, k = 50 : [0.16, 0.86, 0.8]
Précision pour f = 0.15, k = 100 : [0.47, 0.8, 0.76]
Précision pour f = 0.15, k = 200 : [0.54, 0.835, 0.81]
Précision pour f = 0.15, k = 400 : [0.5225, 0.7675, 0.7475]
Précision pour f = 0.2, k = 50 : [0.24, 0.86, 0.84]
Précision pour f = 0.2, k = 100 : [0.48, 0.89, 0.83]
Précision pour f = 0.2, k = 200 : [0.57, 0.84, 0.84]
Précision pour f = 0.2, k = 400 : [0.535, 0.8475, 0.8375]